Add workflow 2 to main.py

PotatoCombat opened the pull request:
Part of #12

PotatoCombat committed:
Add template for workflow 2

PotatoCombat committed:
Fix numbering of options

PotatoCombat committed:
Add option to cancel workflow 2

PotatoCombat committed:
Fix Spotify song database entries

- Replace array entries with CSV
- Add index column label
- Add "simple_name" column to simplify filtering names with ASCII keywords

PotatoCombat committed:
Fix nearest neighbors for workflow 2

PotatoCombat committed:
Fix display of workflow 2 playlists

PotatoCombat committed:
Add display seeds option in workflow 2

PotatoCombat committed:
Format function spacings

PotatoCombat committed:
Remove unused code and comments

PotatoCombat committed:
Merge branch 'workflow2'

PotatoCombat commented:
There might be some minor bugs in our data still. Not sure if there are duplicate entries.

PotatoCombat commented:
Would be nice to have a second opinion about the nearest neighbours results!

PotatoCombat commented:
Repeated nearest neighbours (NN) for each seed in the list.
Let me know if it should be implemented differently, e.g. use the average features of the seeds.

madanalogy commented on main.py:
```diff
@@ -29,7 +34,7 @@ def workflow_select():
     print("[2] Generate a playlist based on seed song(s)")
     workflow_input = input()
-    if workflow_input != "1" or workflow_input != "2":
+    if workflow_input != "1" and workflow_input != "2":
```
Good catch haha

madanalogy commented on main.py:
```diff
@@ -101,9 +106,178 @@ def workflow_1(playlist_length):
+    playlist_songs = []
+    finished_workflow = False
+
+    while not finished_workflow:
```
Could we do validation on the seed length as well? Since I'm assuming the seed songs will be part of the playlist, it shouldn't be longer than `playlist_length`? Maybe:
```suggestion
    while not finished_workflow and len(seeds) < playlist_length - 1:
```

PotatoCombat replied:
Okie. I've added this in the latest commits!

madanalogy commented on playlist_generator/Nearest_Neighbours.py:
```diff
@@ -81,26 +101,3 @@ def find_songs_by_valence(genre, valence, n=10):
     return songs
-
-
-if __name__ == '__main__':
```
I think you could leave this here tbh, will only ever run with the command `python Nearest_Neighbours.py`

PotatoCombat replied:
Haha I changed the behaviour of the functions so this code wouldn't work anymore. I get what you mean though.

madanalogy replied:
Yea no worries, can just remove it, I was just wondering haha

madanalogy reviewed (commented):
Looks good. Why did the data need to change again? I thought there weren't any duplicate song IDs?

madanalogy commented:
> Repeated nearest neighbours (NN) for each seed in the list.

How does this work actually? If I wanted 20 songs in my playlist and give 5 songs as seeds, do I get 100 songs? :")

> Let me know if it should be implemented differently, e.g. use the average features of the seeds.

I think this was what we discussed during the last meeting. Just get an aggregation and use that to populate the nearest neighbours

PotatoCombat committed:
Add aggregation of seed features in workflow 2

PotatoCombat commented:
> > Repeated nearest neighbours (NN) for each seed in the list.
> 
> How does this work actually? If I wanted 20 songs in my playlist and give 5 songs as seeds, do I get 100 songs? :")

Yup 🙈 . Didn't sound right.

> > Let me know if it should be implemented differently, e.g. use the average features of the seeds.
> 
> I think this was what we discussed during the last meeting. Just get an aggregation and use that to populate the nearest neighbours

Okie fixed this. Will update pull request soon.

PotatoCombat committed:
Fix Nearest Neighbours dependencies in main

- Combine display song functions
- Fix data file paths

PotatoCombat committed:
Add seed limit in workflow 2

PotatoCombat committed:
Fix workflow 1 data sets

- Fix comma separaton of array values
- Fix invalid special characters
- Fix index column

PotatoCombat committed:
Refactor file path functions

PotatoCombat committed:
Fix index col of workflow 2 data set

PotatoCombat committed:
Fix workflow 2 data sets

- Fix comma separaton of array values
- Fix invalid special characters
- Fix index column

PotatoCombat committed:
Replace workflow 2 search results structure

Original: Return a list of song names
Modified: Return a list of song numbers (their row index in the data
set)

PotatoCombat committed:
Format pretty printing of option menus

PotatoCombat committed:
Merge branch 'workflow2'

PotatoCombat commented:
Had to do extra cleaning on our data set. Here is an outline of the work performed on the data sets:
- There weren't any duplicates: they were identical songs released in different years/albums.
- Some song names/artists had invalid characters (such as unparsed apostrophes, raw unicode chars)
- Replaced index columns so that numbering starts at 0
- Fixed entries with duplicate artists

PotatoCombat commented on main.py:
```diff
@@ -101,9 +105,205 @@ def workflow_1(playlist_length):
+            continue
+
+        if option == '2':
+            if len(seeds) < seed_limit:
```
Included seed limit here. If above the seed limit, this option cannot be selected

madanalogy reviewed (approved):
LGTM
